docs/swarm_scatter_demo.man
swarm_scatter_demo(1)
=====================

NAME
---- 
swarm_scatter_demo - General demo program for running many n-body integrations.

SYMPOSIS
--------
*swarm_scatter_demo* [ -cfg <cfg directory/filename> ] [ -obs <list-of-times directory/filename> ] [ -ics <IC directory/filename prefix> ] [ -out <output directory/filename prefix> ]
Several parameters must be specified in the configuration file (.cfg). If no configuration file, including the directory, is specified, the cfg file is assumed to be ''$RUN/integrator.cfg'', where $RUN is the directory from which the swarm_scatter_demo is being executed. All options except for -cfg can also be specified in the configuration file, which supersedes the command line.  If no options are given, the default values given in swarm_scatter_demo.h or user.h are used. For this man page, assume that $RUN=$SWARMHOME/run.

Description
-----------
swarm_scatter_demo  -- Runs many n-body integrations with initial conditions generated by easyGen.py, swarm_scatter_demo.py, or throwBinaryGen.py.  An observation file must also be supplied.  This observation file is a single column list of times for which phase space data will be stored. The python scripts will create an observation file automatically when generating ICs (see below), but the file can also be generated separately. The output times need not have constant intervals. The demo code also uses the swarm utilities copy_from and accessor functions for formatting data output. For the configuration file, several parameters must be specified.  Each parameter must appear on its own line with the format: +
''key = value''

REQUIRED PARAMETERS
~~~~~~~~~~~~~~~~~~~

''integrator''::
   Specifies one of the following integrators: ''gpu_hermite'', ''gpu_hermite_adap'', ''gpu_rk4'', ''gpu_rkck'', ''gpu_verlet'', ''cpu_hermite''

''min time step''::
   For adaptive time stepping, ''min time step'' is the minimum time step allowed.

''precision''::
   Specifies the precision for the integrations.  0 results in single precision and 1 in double.  Some integrators support additional values that allow  mixed-precision calculations.  For example, gpu_hermite and gpu_hermite_adap allow a value of 2, which calculates positions and velocities in double precision, but accelerations and jerks are calculated in single precision.

''runon''::
   Specifies either ''gpu'' or ''cpu''.  In the initial release, the values of runon and integrator must match.  In future relases, the CPU and GPU integrators may be merged.

''time step''::
  For the fixed time stepping, ''time step'' is the constant time step. _Not_ requried for gpu_hermite_adap.

''time step factor''::
   For adaptive time stepping, ''time step factor'' is the fraction of the timescale used in determining the time step.  

OPTIONAL PARAMETERS
~~~~~~~~~~~~~~~~~~~

''initial conditions''::
   File prefix, including the directory path, for the ICs.  For example, ''../scripts/ic/data'' will look for files called ''../scripts/ic/data.*'' for * in the range [0,n). Specifying this option will supersede the command line option ''-ics <IC filename prefix>''. The default for the distribution is ''$RUN/../scripts/ic/data''.

''list of times''::
   Filename, including directory path, that contains the list of times for which output will be generated.  Specifying this option will supersede the command line option ''-obs <observation filename>''. The default for the distribution is ''$RUN/../scripts/ic/observeTimes.dat''.

''output'':: 
   Specifies files to be used for built-in GPU logging.  _Only some integrators support this option at this time_. See *eventlog* for more information.

''out files''::
   File prefix, including the directory path, for the output files.  Specifying this option will supersede the command line option ''-ics <output filename prefix>''. The default for the distribution is ''$RUN/scatter_output''.  Note that this directory must be created by the user.

''output interval''::
   Specifies how often data should be logged using the built-in GPU logging. _Only some integrators support this option at this time_. See *eventlog* for more information.

''rmax''::
   Stop integration once a body reaches a distance greater than ''rmax''. _Only some integrators support this option at this time_.

''threads per block''::
   Specifies how many threads the GPU uses per block. _We recommend most users not set this explicitly, except when performing performance benchmarks._  Default value is 64 in the current release, but future releases will likely choose a good value of this parameter automatically.


EXAMPLE
-------

First, make the applications with 'make' and generate files specifying the initial conditions, e.g.
'cd scripts/ic; ./swarm_scatter_demo.py ; cd ../..`
'cd run; mkdir scatter_output; ../bin/swarm_scatter_demo -cfg integrator-scatter.cfg'',
where integrator-scatter.cfg contains the following:

[literal]
-----------------------------------
# swarm_scatter_demo cfg parameters
initial conditions = ../scripts/ic/data
list of times = ../scripts/ic/observeTimes.dat

#integration
integrator = gpu_hermite_adap
runon = gpu

# integrator-specific parameters
min time step = 1e-6
time step factor = 0.03
precision = 1

# GPU kernel configuration
threads per block = 64

# Output parameters
output = null
out files = ./scatter_output/scatter
-----------------------------------


Alternatively, one could use some command line options.  For example, ''../bin/swarm_scatter_demo -cfg integrator-scatter.cfg -ics ../scripts/ic/data -obs ./localObsTimeFile.dat'',
where integrator-scatter.cfg now contains the following:

[literal]
-----------------------------------
# swarm_scatter_demo cfg parameters

#integration
integrator = gpu_hermite_adap
runon = gpu

# integrator-specific parameters
min time step = 1e-6
time step factor = 0.03
precision = 1

# GPU kernel configuration
threads per block = 64

# Output parameters
output = null
out files = ../scatter_output/scatter
-----------------------------------


_NOTE_ that the user must ensure that the ICs have been made available and that the output directory has been created.  The user may choose to dump the output in the $RUN directory, but this could make for a messy directory. _Remember_, configuration file options supersede command line options.

OUTPUT
~~~~~~

The current output for swarm_scatter_demo writes a single line per system observation time, as specified by the list of times file.  The data are ordered as follows: 

[literal]
--------------------------------------------------------------------------
time_1 {for(i=0; i<N-1;++i){m_i x_i y_i z_i vx_i vy_i vz_i }} change in total_energy
time_2 {for(i=0; i<N-1;++i){m_i x_i y_i z_i vx_i vy_i vz_i }} change in total_energy
time_3 {for(i=0; i<N-1;++i){m_i x_i y_i z_i vx_i vy_i vz_i }} change in total_energy
.
.
.
time_N {for(i=0; i<N-1;++i){m_i x_i y_i z_i vx_i vy_i vz_i }} change in total_energy
--------------------------------------------------------------------------


INITIAL CONDITIONS
~~~~~~~~~~~~~~~~~~

The intial conditions can be generated by easyGen.py or throwBinaryGen.py (both in the $SWARMHOME/scripts directory) or swarm_scatter_demo.py (in the $SWARMHOME/scripts/ic directory). The user may wish to write his or her own IC generator.  In this case, the ICs must be formated as follows:

[literal]
---------------------------------
Total_number_of_bodies
Mass_1 x_1 y_1 z_1 vx_1 vy_1 vz_1
Mass_2 x_2 y_2 z_2 vx_2 vy_2 vz_2
Mass_3 x_3 y_3 z_3 vx_3 vy_3 vz_3
.
.
. 
Mass_N x_N y_N z_N vx_N vy_N vz_N
---------------------------------


LIMITATIONS & BUGS
------------------
See http://www.astro.ufl.edu/~eford/code/swarm/[website] for updated information on bugs.

RESOURCES
---------


Main Web site: <http://www.astro.ufl.edu/~eford/code/swarm/>

Google Group: <http://groups.google.com/group/swarm-ng>

COPYING
-------
Copyright \(C) 2008-2010 A. C. Boley + 
Free use of this software is granted under the terms of the GNU
General Public License (GPL) v3, which should be included as
link:gpl-3.0.txt[] and also avaliable
http://www.gnu.org/licenses/gpl-3.0.html[online].



scripts/ic/swarm_scatter_demo.py
#!/usr/bin/python
''' Simple python script for generating an ensemble of planetary systems with an intruding star.'''
#
#    "swarm_scatter_demo.py" is a python script that creates initial conditions and the observation
#    file for use in swarm_scatter_demo.
#    "swarm_scatter_demo" is a program that uses the Swarm-NG tools for modeling an ensemble of
#    small N systems using the hermite_adap_gpu integrator.
#    Copyright (C) 2010  Swarm-NG Development Group
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Contact aaron.boley at the domain gmail.com if you questions regarding
#    this software.
#
#
import math as M
import random as R
import scipy.special as SS
#
#
# define basic parameters for ensemble
# keep the primary mass the same for now
# planets can vary in mass, size, location, and number
# this does not consider stability, except for the Hill radius check
# the outer planet is always set to 128 AU, but that can be changed where indicated
# perturbers are drawn from a Maxwellian velocity distribution
#
# 
nSystems=1000 # systems in ensemble
mPrimary=1. # mass of the primary
nOther=2 # primary plus perturbers. E.g., 3 = primary plus two perturbers
massMin=.001/32. # 10 Earth-mass minimum
massMax=.001 # 1 Jupiter-mass maximum
minPlanets=4 # keeps these the same for now.
maxPlanets=4
minAU=2. # minimum semi-major axis allowed.  If you are running with fixed time steps, be mindful of this setting
maxAU=100. # outer planet
SET_OUTER_ORBIT=1 # force one planet to start at maxAU.
pert=0.01 # perturbations for other velocities.
HILLS=3. # make sure the planets are separated by HILLS many Hill radii
timeStart=0.
timeEnd=1e3 # time should be given in yr.
incomingR=20626.5 # AU
maxUnperturbedImpact=1000. # AU
numObs=100 # number of observations allowed
ObserveFile="observeTimes.dat"
MAXWELL=1 # use max
VelSig=1./29.8 #(1km/s in code units) This is mean of Maxwellian velocity dispersion
ObserveFile="observeTimes.dat" # list of times file used by swarm_scattering_demo
MAXWELL=1 # use Maxwellian velocity.  Anything else gives the same initial speed 
#         # between the binary and the planetary system for each ensemble constituent.
RANDOM_TIMES=0 # set to 1 if you want the list of times file to have random intervals
thisSeed=314159 # random number generator seed


def getUniformLog(b0,b1):
        a0=M.log10(b0)
        a1=M.log10(b1)
        a=10.**(R.uniform(a0,a1))
        return a

def Maxwell(v):
        p=R.random()
        s=0.
        SqrtPi=M.sqrt(M.pi)
        xinit=.001
        x=xinit
        while True:
                s=M.pi/v*(SqrtPi*v*SS.erf(x/v)-2*x*M.exp(-(x/v)**2))/M.pi**1.5
                err=(s-p)/p
                if M.fabs(err)<1e-6:break
                if(err<0.):x=x-x*err*.1
                else:x=x-x*err*.1
                if x<0.:
                        xinit/=2.
                        x=xinit
        return x

def createObservingFile():
        R.seed()
        obsTimes=[]
        obsTimes.append(timeStart)
        if RANDOM_TIMES:
                for i in xrange(1,numObs):
                        obsTimes.append(R.uniform(timeStart,timeEnd))
                obsTimes.sort()
        else:
                dt=(timeEnd-timeStart)/float(numObs)
                for i in xrange(1,numObs+1):
                        obsTimes.append(timeStart+i*dt)

        f=open(ObserveFile,"w")
        for i in xrange(numObs+1):
                f.write(repr(obsTimes[i])+"\n")
        f.close()
        return 0

def getCollision():
	# first use spherical coordinates to find position in sky as observed by planetary system
	R.seed()
	phi=2.*M.pi*R.random()
	theta=M.pi*(2.*R.random()-1.)
        if MAXWELL==1: VelSigPert=Maxwell(VelSig) 
	else: VelSigPert=VelSig
	x=incomingR*M.cos(theta)*M.cos(phi)
	y=incomingR*M.cos(theta)*M.sin(phi)
	z=incomingR*M.sin(theta)
        impact=M.sqrt(maxUnperturbedImpact**2+4.*mPrimary*maxUnperturbedImpact/VelSigPert**2)  # assumes perturber is the same mass as the primary
	impact*=M.sqrt(R.random())
	if impact>incomingR:impact=incomingR
	alpha0=impact/incomingR
	alpha=1e-3*alpha0 # take advantage of large separation
	ppf=(2.*(R.random())-1.)*alpha
	tpf=(2.*(R.random())-1.)*alpha
	while alpha<alpha0:
       		phiPrime=phi+ppf
               	thetaPrime=theta+tpf
		ppf*=1.01;tpf*=1.01
		alpha=M.cos(theta)*M.cos(phi)*M.cos(thetaPrime)*M.cos(phiPrime)
		alpha+=M.cos(theta)*M.sin(phi)*M.cos(thetaPrime)*M.sin(phiPrime)
		alpha+=M.sin(thetaPrime)*M.sin(theta)
		alpha=M.acos(alpha)
	vx=-VelSigPert*M.cos(thetaPrime)*M.cos(phiPrime)
	vy=-VelSigPert*M.cos(thetaPrime)*M.sin(phiPrime)
	vz=-VelSigPert*M.sin(thetaPrime)
	return x,y,z,vx,vy,vz

def main():
	R.seed()
	for i in xrange(nSystems):
		nPlanets=R.randint(minPlanets,maxPlanets)
		print " Working on system ", i
		buffer="data."+repr(i)
		f=open(buffer,"w")

		f.write(repr(nPlanets+nOther)+"\n")
                buffer=""
		listx=[]
		pvx=0.;pvy=0.;pvz=0.
		for j in xrange(nPlanets):
			mass=getUniformLog(massMin,massMax)
			x=getUniformLog(minAU,maxAU)
			if SET_OUTER_ORBIT==1: 
				if j==0: x=maxAU
			#if j==1: x=2.
			#if j==2: x=8. # set these if you want to constrain planet locations
			#if j==3: x=32.
			listx.append(x)
			OK=0
			if j==0: OK=1
                	while not OK:
				OK=1
				for k in xrange(j):
					minSep=abs(x)*HILLS*(mass/mPrimary/3.)**(1./3.)
					if( abs(abs(x)-abs(listx[k]))<minSep):
						OK=0
						x=getUniformLog(minAU,maxAU)
						listx[j]=x
					

			y=0.;z=0.
			vy=x/abs(x)*M.sqrt(mPrimary/abs(x))
			vx=vy*pert*(R.random()*2.-1.)
			vz=vy*pert*(R.random()*2.-1.)
                        pvy-=vy*mass
			pvx-=vx*mass
			pvz-=vz*mass
			buffer+=repr(mass)+" "+repr(x)+" "+repr(y)+" "+repr(z)+" "+repr(vx)+" "+repr(vy)+" "+repr(vz)+"\n"
		buffer0=repr(mPrimary)+" "+repr(0.)+" "+repr(0.)+" "+repr(0.)+" "+repr(pvx/mPrimary)+" "+repr(pvy/mPrimary)+" "+repr(pvz/mPrimary)+"\n"
		f.write(buffer0)
		f.write(buffer)
		
		# now find and write perturber information
		for j in xrange(nOther-1):
			x,y,z,vx,vy,vz=getCollision()
			buffer=repr(mPrimary)+" "+repr(x)+" "+repr(y)+" "+repr(z)+" "+repr(vx)+" "+repr(vy)+" "+repr(vz)+"\n"
			f.write(buffer)

		f.close()		

	test=createObservingFile()
        if test!=0:
		print "Error when creating observing file."
		return -1
 
	return 0

test=main()
if test==0: print "Program exited sucessfully."
else: "Error in main. Results may be unreliable."
